File: app/src/main/java/closer/vlllage/com/closer/handler/data/PersistenceHandler.java
package closer.vlllage.com.closer.handler.data;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import com.google.android.gms.maps.model.LatLng;

import closer.vlllage.com.closer.handler.helpers.ApplicationHandler;
import closer.vlllage.com.closer.pool.PoolMember;

public class PersistenceHandler extends PoolMember {

    private static final String SHARED_PREFERENCES = "closer.prefs";
    private static final String PREFERENCE_MY_STATUS = "closer.me.status";
    private static final String PREFERENCE_MY_NAME = "closer.me.name";
    private static final String PREFERENCE_MY_ACTIVE = "closer.me.active";
    private static final String PREFERENCE_MY_PHOTO = "closer.me.photo";
    private static final String PREFERENCE_MY_PRIVATE_MODE = "closer.me.private-mode";
    private static final String PREFERENCE_DEVICE_TOKEN = "closer.device-token";
    private static final String PREFERENCE_PHONE = "closer.message";
    private static final String PREFERENCE_VERIFIED = "closer.verified";
    private static final String PREFERENCE_PHONE_ID = "closer.message.id";
    private static final String PREFERENCE_HELP_IS_HIDDEN = "closer.state.help-is-hidden";
    private static final String PREFERENCE_NOTIFICATIONS_PAUSED = "closer.notifications.paused";
    private static final String PREFERENCE_LAST_MAP_CENTER = "closer.map.center";

    private SharedPreferences sharedPreferences;

    @Override
    protected void onPoolInit() {
        sharedPreferences = $(ApplicationHandler.class).getApp().getSharedPreferences(
                SHARED_PREFERENCES, Context.MODE_PRIVATE
        );
    }

    public String getMyStatus() {
        return sharedPreferences.getString(PREFERENCE_MY_STATUS, "");
    }

    public String getMyName() {
        return sharedPreferences.getString(PREFERENCE_MY_NAME, "");
    }

    public String getMyPhoto() {
        return sharedPreferences.getString(PREFERENCE_MY_PHOTO, "");
    }

    public boolean getMyActive() {
        return sharedPreferences.getBoolean(PREFERENCE_MY_ACTIVE, false);
    }

    public String getDeviceToken() {
        return sharedPreferences.getString(PREFERENCE_DEVICE_TOKEN, null);
    }

    public String getPhone() {
        return sharedPreferences.getString(PREFERENCE_PHONE, null);
    }

    public boolean getIsVerified() {
        return sharedPreferences.getBoolean(PREFERENCE_VERIFIED, false);
    }

    public String getPhoneId() {
        return sharedPreferences.getString(PREFERENCE_PHONE_ID, null);
    }

    public boolean getIsHelpHidden() {
        return sharedPreferences.getBoolean(PREFERENCE_HELP_IS_HIDDEN, false);
    }

    public boolean getIsNotificationsPaused() {
        return sharedPreferences.getBoolean(PREFERENCE_NOTIFICATIONS_PAUSED, false);
    }

    public boolean getPrivateMode() {
        return sharedPreferences.getBoolean(PREFERENCE_MY_PRIVATE_MODE, false);
    }

    public LatLng getLastMapCenter() {
        String result = sharedPreferences.getString(PREFERENCE_LAST_MAP_CENTER, null);

        if (result == null) {
            return null;
        }


        String[] latLng = result.split(",");

        return new LatLng(
                Double.valueOf(latLng[0]),
                Double.valueOf(latLng[1])
        );
    }

    @SuppressLint("ApplySharedPref")
    public void setMyStatus(String status) {
        sharedPreferences.edit().putString(PREFERENCE_MY_STATUS, status).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setMyName(String name) {
        sharedPreferences.edit().putString(PREFERENCE_MY_NAME, name).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setMyPhoto(String name) {
        sharedPreferences.edit().putString(PREFERENCE_MY_PHOTO, name).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setMyActive(boolean active) {
        sharedPreferences.edit().putBoolean(PREFERENCE_MY_ACTIVE, active).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setDeviceToken(String deviceToken) {
        sharedPreferences.edit().putString(PREFERENCE_DEVICE_TOKEN, deviceToken).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setPhone(String phone) {
        sharedPreferences.edit().putString(PREFERENCE_PHONE, phone).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setIsVerified(boolean verified) {
        sharedPreferences.edit().putBoolean(PREFERENCE_VERIFIED, verified).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setPhoneId(String phoneId) {
        sharedPreferences.edit().putString(PREFERENCE_PHONE_ID, phoneId).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setIsHelpHidden(boolean helpIsHidden) {
        sharedPreferences.edit().putBoolean(PREFERENCE_HELP_IS_HIDDEN, helpIsHidden).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setIsNotificationsPaused(boolean isNotificationsPaused) {
        sharedPreferences.edit().putBoolean(PREFERENCE_NOTIFICATIONS_PAUSED, isNotificationsPaused).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setPrivateMode(boolean privateMode) {
        sharedPreferences.edit().putBoolean(PREFERENCE_MY_PRIVATE_MODE, privateMode).commit();
    }

    @SuppressLint("ApplySharedPref")
    public void setLastMapCenter(LatLng latLng) {
        sharedPreferences.edit().putString(PREFERENCE_LAST_MAP_CENTER, latLng.latitude + "," + latLng.longitude).commit();
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/data/PersistenceHandler.kt
package closer.vlllage.com.closer.handler.data

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import closer.vlllage.com.closer.ContentViewType
import closer.vlllage.com.closer.handler.feed.FeedContent
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler
import closer.vlllage.com.closer.handler.helpers.Val
import com.google.android.gms.maps.model.LatLng
import com.queatz.on.On
import com.queatz.on.OnLifecycle
import io.reactivex.subjects.PublishSubject

class PersistenceHandler constructor(private val on: On) : OnLifecycle {

    private val listener = SharedPreferences.OnSharedPreferenceChangeListener { sharedPreferences, key ->
        Companion.changes.onNext(key)
    }

    private lateinit var sharedPreferences: SharedPreferences

    val changes = Companion.changes

    var appsToolbarOrder: List<ContentViewType>
        get() = sharedPreferences.getString(PREFERENCE_APPS_TOOLBAR_ORDER, null)?.split(",")?.mapNotNull {
            try {
                ContentViewType.valueOf(it)
            } catch (e: Exception) {
                null
            }
        } ?: listOf()
        @SuppressLint("ApplySharedPref")
        set(value) {
            sharedPreferences.edit().putString(PREFERENCE_APPS_TOOLBAR_ORDER, value.joinToString(",") { it.name }).commit()
        }

    var myStatus: String
        get() = sharedPreferences.getString(PREFERENCE_MY_STATUS, "")!!
        @SuppressLint("ApplySharedPref")
        set(status) {
            sharedPreferences.edit().putString(PREFERENCE_MY_STATUS, status).commit()
        }

    var myName: String
        get() = sharedPreferences.getString(PREFERENCE_MY_NAME, "")!!
        @SuppressLint("ApplySharedPref")
        set(name) {
            sharedPreferences.edit().putString(PREFERENCE_MY_NAME, name).commit()
        }

    var myPhoto: String
        get() = sharedPreferences.getString(PREFERENCE_MY_PHOTO, "")!!
        @SuppressLint("ApplySharedPref")
        set(name) {
            sharedPreferences.edit().putString(PREFERENCE_MY_PHOTO, name).commit()
        }

    var myActive: Boolean
        get() = sharedPreferences.getBoolean(PREFERENCE_MY_ACTIVE, false)
        @SuppressLint("ApplySharedPref")
        set(active) {
            sharedPreferences.edit().putBoolean(PREFERENCE_MY_ACTIVE, active).commit()
        }

    var access: Boolean
        get() = sharedPreferences.getBoolean(PREFERENCE_ACCESS, false)
        @SuppressLint("ApplySharedPref")
        set(active) {
            sharedPreferences.edit().putBoolean(PREFERENCE_ACCESS, active).commit()
        }

    var deviceToken: String?
        get() = sharedPreferences.getString(PREFERENCE_DEVICE_TOKEN, null)
        @SuppressLint("ApplySharedPref")
        set(deviceToken) {
            sharedPreferences.edit().putString(PREFERENCE_DEVICE_TOKEN, deviceToken).commit()
        }

    var phone: String?
        get() = sharedPreferences.getString(PREFERENCE_PHONE, null)
        @SuppressLint("ApplySharedPref")
        set(phone) {
            sharedPreferences.edit().putString(PREFERENCE_PHONE, phone).commit()
        }

    var isVerified: Boolean
        get() = sharedPreferences.getBoolean(PREFERENCE_VERIFIED, false)
        @SuppressLint("ApplySharedPref")
        set(verified) {
            sharedPreferences.edit().putBoolean(PREFERENCE_VERIFIED, verified).commit()
        }

    var phoneId: String?
        get() = sharedPreferences.getString(PREFERENCE_PHONE_ID, null)
        @SuppressLint("ApplySharedPref")
        set(phoneId) {
            sharedPreferences.edit().putString(PREFERENCE_PHONE_ID, phoneId).commit()
        }

    var isNotificationsPaused: Boolean
        get() = sharedPreferences.getBoolean(PREFERENCE_NOTIFICATIONS_PAUSED, false)
        @SuppressLint("ApplySharedPref")
        set(isNotificationsPaused) {
            sharedPreferences.edit().putBoolean(PREFERENCE_NOTIFICATIONS_PAUSED, isNotificationsPaused).commit()
        }

    var privateMode: Boolean
        get() = sharedPreferences.getBoolean(PREFERENCE_MY_PRIVATE_MODE, false)
        @SuppressLint("ApplySharedPref")
        set(privateMode) {
            sharedPreferences.edit().putBoolean(PREFERENCE_MY_PRIVATE_MODE, privateMode).commit()
        }

    var privateOnly: Boolean
        get() = sharedPreferences.getBoolean(PREFERENCE_MY_PRIVATE_ONLY, false)
        @SuppressLint("ApplySharedPref")
        set(privateOnly) {
            sharedPreferences.edit().putBoolean(PREFERENCE_MY_PRIVATE_ONLY, privateOnly).commit()
        }

    var lastMapCenter: LatLng?
        get() {
            val result = sharedPreferences.getString(PREFERENCE_LAST_MAP_CENTER, null)
                    ?: return null

            val latLng = result.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            return LatLng(
                    java.lang.Double.valueOf(latLng[0]),
                    java.lang.Double.valueOf(latLng[1])
            )
        }
        @SuppressLint("ApplySharedPref")
        set(latLng) {
            sharedPreferences.edit().putString(PREFERENCE_LAST_MAP_CENTER, latLng!!.latitude.toString() + "," + latLng.longitude).commit()
        }

    var lastFeedTab: FeedContent?
        get() = sharedPreferences.getString(PREFERENCE_LAST_FEED_CONTENT, null)?.let { on<Val>().valueOr(it, FeedContent.WELCOME) }
        @SuppressLint("ApplySharedPref")
        set(feedContent) {
            sharedPreferences.edit().putString(PREFERENCE_LAST_FEED_CONTENT, feedContent?.name).commit()
        }

    override fun on() {
        sharedPreferences = on<ApplicationHandler>().app.getSharedPreferences(
                SHARED_PREFERENCES, Context.MODE_PRIVATE
        )

        sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
    }

    companion object {
        val changes = PublishSubject.create<String>()

        const val SHARED_PREFERENCES = "closer.prefs"
        const val PREFERENCE_APPS_TOOLBAR_ORDER = "closer.apps.toolbar.order"
        const val PREFERENCE_MY_STATUS = "closer.me.status"
        const val PREFERENCE_MY_NAME = "closer.me.name"
        const val PREFERENCE_MY_ACTIVE = "closer.me.active"
        const val PREFERENCE_MY_PHOTO = "closer.me.photo"
        const val PREFERENCE_MY_PRIVATE_MODE = "closer.me.private-mode"
        const val PREFERENCE_MY_PRIVATE_ONLY = "closer.me.private-only"
        const val PREFERENCE_DEVICE_TOKEN = "closer.device-token"
        const val PREFERENCE_PHONE = "closer.phone"
        const val PREFERENCE_VERIFIED = "closer.verified"
        const val PREFERENCE_PHONE_ID = "closer.phone.id"
        const val PREFERENCE_NOTIFICATIONS_PAUSED = "closer.notifications.paused"
        const val PREFERENCE_LAST_MAP_CENTER = "closer.map.center"
        const val PREFERENCE_LAST_FEED_CONTENT = "closer.feed.content"
        const val PREFERENCE_ACCESS = "closer.access"
    }
}
